evalml/tests/component_tests/decomposer_tests/test_decomposer.py

```diff
@@ -481,3 +500,125 @@ def test_decomposer_bad_target_index(
         match="doesn't support target data with index of type",
     ):
         dec._choose_proper_index(y)
+
+
+@pytest.mark.parametrize(
+    "decomposer_child_class",
+    decomposer_list,
+)
+@pytest.mark.parametrize(
+    "transformer_fit_on_data",
+    [
+        "in-sample",
+        "in-sample-less-than-sample",
+        "wholly-out-of-sample",
+        "wholly-out-of-sample-no-gap",
+        "partially-out-of-sample",
+        "out-of-sample-in-past",
+        "partially-out-of-sample-in-past",
+    ],
+)
+def test_decomposer_fit_transform_out_of_sample(
+    decomposer_child_class,
+    generate_seasonal_data,
+    transformer_fit_on_data,
+):
+    # Generate 10 periods (the default) of synthetic seasonal data
+    seasonal_period = 7
+    X, y = generate_seasonal_data(real_or_synthetic="synthetic")(
+        period=seasonal_period,
+        freq_str="D",
+        set_time_index=True,
+        seasonal_scale=0.05,  # Increasing this value causes the decomposer to miscalculate trend
+    )
+    subset_X = X[2 * seasonal_period : 7 * seasonal_period]
+    subset_y = y[2 * seasonal_period : 7 * seasonal_period]
+
+    decomposer = decomposer_child_class(seasonal_period=seasonal_period)
+    decomposer.fit(subset_X, subset_y)
+
+    if transformer_fit_on_data == "in-sample":
+        output_X, output_y = decomposer.transform(subset_X, subset_y)
+        pd.testing.assert_series_equal(
+            pd.Series(np.zeros(len(output_y))).set_axis(subset_y.index),
+            output_y,
+            check_dtype=False,
+            check_names=False,
+            atol=0.2,
+        )
+
+    if transformer_fit_on_data != "in-sample":
+        y_new = build_test_target(
+            subset_y,
+            seasonal_period,
+            transformer_fit_on_data,
+            to_test="transform",
+        )
+        if transformer_fit_on_data in [
+            "out-of-sample-in-past",
+            "partially-out-of-sample-in-past",
+        ]:
+            with pytest.raises(
+                ValueError,
+                match="STLDecomposer cannot transform/inverse transform data out of sample",
+            ):
+                output_X, output_inverse_y = decomposer.transform(None, y_new)
+        else:
+            output_X, output_y_t = decomposer.transform(None, y[y_new.index])
+
+            pd.testing.assert_series_equal(
+                pd.Series(np.zeros(len(output_y_t))).set_axis(y_new.index),
+                output_y_t,
+                check_exact=False,
+                atol=0.1,
```

Comment:
I did have to loosen this tolerance (`atol`) here a bit.  Part of the problem was that the different decomposers are able to more or less properly capture the trend.  Based on how much of the trend they capture, the answers will be closer or further away from a horizontal line.  Maybe this isn't a good way to test the transform...you tell me!